dynamic_programming/knuth_yao_speedup.py

```diff
@@ -0,0 +1,61 @@
+"""
+Knuth-Yao dynamic programming speedup,
+also known as the Knuth-Yao optimization, is a technique that accelerates
+the computation of certain dynamic programming algorithms.
+It was introduced by Donald Knuth and Andrew Yao in their paper
+"Efficient Binary-Search Trees" in 1976.
+
+This implementation the following recurrence relation,
+
+dp[i][j] = min(i < k < j){dp[i][k] + dp[k][j] + cost[i][j]},
+
+where opt[i][j-1] <= opt[i][j] <= opt[i+1][j] with opt[i][j] representing
+the value of 'k' that minimizes the given expression.
+
+Equivalently, the cost function satisfies either of the following conditions.
+
+1) cost[b][c] <= cost[a][d]
+2) cost[a][c]+cost[b][d] <= cost[a][d]+cost[b][c]
+
+Reference: https://cp-algorithms.com/dynamic_programming/knuth-optimization.html
+
+- time complexity: O(n^2)
+- space complexity: O(n^2)
+
+>>> knuth_yao_speedup([[1,2,3,4],[3,4,5,1],[1,1,1,3],[2,2,2,2]])
+15
+"""
+
+from __future__ import annotations
+
+import sys
+
+
+def knuth_yao_speedup(cost: list[list[int]]) -> int:
+    n = len(cost)
```

Comment:
add below doctests or add yours if you have...

        >>> knuth_yao_speedup([[1]])
        0
        >>> knuth_yao_speedup([[1, 2], [3, 4]])
        5
        >>> knuth_yao_speedup([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
        15

Reply:
My implementation intended to solve the recurrence relation of 
`dp[i][j] = min(i < k < j){dp[i][k] + dp[k][j] + cost[i][j]}` which can solve this [cutting stick problem](https://cs.stackexchange.com/questions/59949/knuth-yao-dp-speedup-cutting-sticks).

With this recurrence relation, outputs are

    >>> knuth_yao_speedup([[1, 2], [3, 4]])
    2
    >>> knuth_yao_speedup([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
    11

as 

in case of cost function is [[1,2],[3,4]], 
dp[0][1] == cost[0][1] == 2

and in case of cost function is [[1, 2, 3], [4, 5, 6], [7, 8, 9]],
dp[0][2] == dp[0][1]+dp[1][2]+cost[0][2] == cost[0][1]+cost[1][2]+cost[0][2] == 2 + 6 + 3 == 11.

Reply:
Ahh! My mistake....thanks for the correction.

Reply:
Added your examples too!